enterprise/server/backends/redis_execution_collector/redis_execution_collector.go

```diff
@@ -294,11 +294,19 @@ func unmarshalStoredInvocationLinks(serializedResults []string) ([]*sipb.StoredI
 // (exit code, stats, etc.)
 func mergeExecutionUpdates(serializedResults []string) (*repb.StoredExecution, error) {
 	out := &repb.StoredExecution{}
+	lastStage := int64(repb.ExecutionStage_UNKNOWN)
 	for _, serializedResult := range serializedResults {
 		event := &repb.StoredExecution{}
 		if err := proto.Unmarshal([]byte(serializedResult), event); err != nil {
 			return nil, err
 		}
+
+		// The scheduler attempts to prevent concurrent execution updates via
+		// task leasing, but this is not currently 100% reliable. If execution
+		// progress appears to restart, ignore any future updates.
+		if lastStage == int64(repb.ExecutionStage_COMPLETED) && event.GetStage() != int64(repb.ExecutionStage_COMPLETED) {
+			break
+		}
```

Comment:
FYI we might (rarely) see some sequences like

(attempt 1) COMPLETED (execution completed)
(attempt 2) EXECUTING
(attempt 1) COMPLETED (recycling completed)

So we would drop the second COMPLETED event in this case. But it probably happens rarely enough that it's not worth worrying about for now.